drivers/pwm/pwm_max31790.c

```diff
@@ -0,0 +1,703 @@
+						 value_tach_target_count_lsb);
+
+		if (result != 0) {
+			return result;
+		}
+	}
+
+	result = max31790_write_register(dev, register_fan_configuration, value_fan_configuration);
+
+	if (result != 0) {
+		return result;
+	}
+
+	result = max31790_write_register(dev, register_fan_dynamics, value_fan_dynamics);
+
+	if (result != 0) {
+		return result;
+	}
+
+	return 0;
+}
+
+static int max31790_set_cycles(const struct device *dev, uint32_t channel, uint32_t period_count,
+			       uint32_t pulse_count, pwm_flags_t flags)
+{
+	const struct max31790_config *config = dev->config;
+	struct max31790_data *data = dev->data;
+	int result;
+
+	LOG_DBG("set period %i with pulse %i for channel %i and flags 0x%04X", period_count,
+		pulse_count, channel, flags);
+
+	if (channel > config->channel_count) {
+		LOG_ERR("invalid channel number %i", channel);
+		return -EINVAL;
+	}
+
+	if (period_count == 0) {
+		LOG_ERR("period count must be > 0");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+	result = max31790_set_cycles_internal(dev, channel, period_count, pulse_count, flags);
+	k_mutex_unlock(&data->lock);
+	return result;
+}
+
+static int max31790_get_cycles_per_sec(const struct device *dev, uint32_t channel, uint64_t *cycles)
+{
+	const struct max31790_config *config = dev->config;
+	struct max31790_data *data = dev->data;
+	int result;
+	uint8_t pwm_frequency_register;
+	uint8_t pwm_frequency = 1;
+	uint16_t pwm_frequency_in_hz;
+
+	if (channel > config->channel_count) {
+		LOG_ERR("invalid channel number %i", channel);
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+	result = max31790_read_register(dev, MAX37190_GLOBALCONFIGURATION, &pwm_frequency_register);
+
+	if (result == 0) {
+		switch (channel) {
+		case 0:
+		case 1:
+		case 2:
+			pwm_frequency = MAX37190_PWMFREQUENCY_PWM1TO3_GET(pwm_frequency_register);
+			break;
+		case 3:
+		case 4:
+		case 5:
+			pwm_frequency = MAX37190_PWMFREQUENCY_PWM4TO6_GET(pwm_frequency_register);
+			break;
+		}
+
+		if (max31790_convert_pwm_frequency_into_hz(&pwm_frequency_in_hz, pwm_frequency)) {
+			*cycles = pwm_frequency_in_hz;
+		} else {
+			result = -EINVAL;
+		}
+	}
+	k_mutex_unlock(&data->lock);
+
+	return result;
+}
+
+static const struct pwm_driver_api max31790_api = {
+	.set_cycles = max31790_set_cycles,
+	.get_cycles_per_sec = max31790_get_cycles_per_sec,
+};
+
+static int max31790_init(const struct device *dev)
+{
+	const struct max31790_config *config = dev->config;
+	struct max31790_data *data = dev->data;
+	int result;
+	uint8_t register_value;
+
+	k_mutex_init(&data->lock);
+
+	if (!i2c_is_ready_dt(&config->i2c)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	}
+
+	result = max31790_read_register(dev, MAX37190_GLOBALCONFIGURATION, &register_value);
+
+	if (result != 0) {
+		return result;
+	}
+
+	if (MAX37190_GLOBALCONFIGURATION_STANDBY_GET(register_value) == 0b1) {
+		LOG_ERR("PWM controller is in standby");
+		return -ENODEV;
+	}
```

Comment:
I take it there is no way to take the controller out of stand-by mode programmatically?

Reply:
Yes, there is. Actually the same bit is R/W, hence clearing this bit would put the device in standby. By default it will be in run mode after being powered. I put this just for the purpose of checking if the device is correctly configured and accessible. If you mind it I can also remove the register read operation and the check altogether.

Reply:
What if a bootloader or a previous software instance placed the device in standby? The driver needs to take it out of standby in the init function.

Reply:
done